data-prepper-plugins/saas-source-plugins/saas-crawler/src/main/java/org/opensearch/dataprepper/plugins/source/saas/crawler/base/Crawler.java

```diff
@@ -0,0 +1,85 @@
+package org.opensearch.dataprepper.plugins.source.saas.crawler.base;
+
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.model.source.coordinator.enhanced.EnhancedSourceCoordinator;
+import org.opensearch.dataprepper.plugins.source.saas.crawler.coordination.partition.SaasSourcePartition;
+import org.opensearch.dataprepper.plugins.source.saas.crawler.coordination.state.SaasWorkerProgressState;
+import org.opensearch.dataprepper.plugins.source.saas.crawler.model.ItemInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Named;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Named
+public class Crawler {
+    private static final Logger log = LoggerFactory.getLogger(Crawler.class);
+    private static final int maxItemsPerPage = 20;
+    public static final String CREATED = "created";
+    public static final String UPDATED = "updated";
+
+    private final SaasClient client;
+
+    public Crawler(SaasClient client) {
+        this.client = client;
+    }
+
+    public long crawl(long lastPollTime,
+                      EnhancedSourceCoordinator coordinator) {
+        long startTime = System.currentTimeMillis();
+        client.setLastPollTime(lastPollTime);
+        Iterator<ItemInfo> itemInfoIterator = client.listItems();
+        log.info("Starting to crawl the source");
+        long updatedPollTime = 0;
+        log.info("Creating Partitions");
+        do {
+            final List<ItemInfo> itemInfoList = new ArrayList<>();
+            for (int i = 0; i < maxItemsPerPage && itemInfoIterator.hasNext(); i++) {
+                ItemInfo nextItem = itemInfoIterator.next();
+                if(nextItem==null) {
+                    //we don't expect null items, but just in case, we'll skip them
+                    continue;
+                }
+                itemInfoList.add(nextItem);
+                Map<String, String> metadata = nextItem.getMetadata();
+                long niCreated = Long.parseLong(metadata.get(CREATED)!=null? metadata.get(CREATED):"0");
+                long niUpdated = Long.parseLong(metadata.get(UPDATED)!=null? metadata.get(UPDATED):"0");
+                updatedPollTime = Math.max(updatedPollTime, niCreated);
+                updatedPollTime = Math.max(updatedPollTime, niUpdated);
+                log.info("updated poll time {}", updatedPollTime);
+            }
+            createPartition(itemInfoList, coordinator);
```

Comment:
Why are we passing a full list to this method? Can we just create each partition inline or do they all need to be stored first?

Reply:
We are passing `maxItemsPerPage` number of items to this method. i.e. the page size in our paginated crawling. All the items in this page will go into one partition (or a work item). Like a partition per page.